### PR TITLE
refactor(device): Make set desired property update method thread safe

### PR DESCRIPTION
**Pull request by bikamani**
+ Make set desired property method thread safe
+ some logging statements
+ documentation changes.

### CODE REVIEW

**Review by vinagesh — APPROVED** [review]

File: iothub/device/src/InternalClient.cs [hunk 1]
```diff
@@ -1102,6 +1103,15 @@ public async Task SetDesiredPropertyUpdateCallbackAsync(DesiredPropertyUpdateCal
         {
             // Codes_SRS_DEVICECLIENT_18_003: `SetDesiredPropertyUpdateCallbackAsync` shall call the transport to register for PATCHes on it's first call.
             // Codes_SRS_DEVICECLIENT_18_004: `SetDesiredPropertyUpdateCallbackAsync` shall not call the transport to register for PATCHes on subsequent calls
+
+            // Wait to acquire the _twinSemaphore. This ensures that concurrently invoked SetDesiredPropertyUpdateCallbackAsync calls are invoked in a thread-safe manner.
+            if (Logging.IsEnabled)
```
Comment (barustum) [inline iothub/device/src/InternalClient.cs:1108]: based on @abhipsaMisra  latest PR, you don't need to do this. Not sure if she merged it in yet or not.

File: iothub/device/src/InternalClient.cs [hunk 2]
```diff
@@ -1122,6 +1132,30 @@ public async Task SetDesiredPropertyUpdateCallbackAsync(DesiredPropertyUpdateCal
...
+                try
+                {
+                    _twinSemaphore.Release();
+                }
+                catch (SemaphoreFullException)
+                {
+                    // this semaphore is typically grabbed at the start of the method, but if
+                    // this method is canceled while waiting to grab the semaphore, then this semaphore.release
+                    // will throw this SemaphoreFullException since it was never grabbed in the first place
+                    if (Logging.IsEnabled)
```
Comment (azabbasi) [inline iothub/device/src/InternalClient.cs:1146]: As @barustum mentioned, you don't need these checks anymore.

**Review by azabbasi — APPROVED** [review]
Just remove the 
`if (Logging.IsEnabled) `
checks and it looks good.

File: iothub/device/src/InternalClient.cs [hunk 3]
```diff
@@ -109,6 +109,7 @@ internal class InternalClient : IDisposable
         private readonly SemaphoreSlim _methodsDictionarySemaphore = new SemaphoreSlim(1, 1);
         private readonly SemaphoreSlim _deviceReceiveMessageSemaphore = new SemaphoreSlim(1, 1);
         private readonly SemaphoreSlim _moduleReceiveMessageSemaphore = new SemaphoreSlim(1, 1);
+        private readonly SemaphoreSlim _twinSemaphore = new SemaphoreSlim(1, 1);
```
Comment (abhipsaMisra) [inline iothub/device/src/InternalClient.cs:112]: nit: rename to twinDesiredPropertySemaphore? To call out that it is relevant only for the desired property flow, and not for reported property/ get twin flow.

File: iothub/device/src/InternalClient.cs [hunk 4]
```diff
@@ -1063,7 +1064,7 @@ private Task EnableMethodAsync(CancellationToken cancellationToken)
         }
 
         /// <summary>
-        /// Sets or clears a callback that will be called whenever the client receives a state update
+        /// Sets a callback that will be called whenever the client receives a state update
```
Comment (abhipsaMisra) [inline iothub/device/src/InternalClient.cs:1067]: I think the previous description was good -> we called out that "set to null" = clear. What do you think?

**Review by abhipsaMisra — APPROVED** [review]
The change looks good, minor comment about renaming and removing logging enabled checks.

File: iothub/device/src/InternalClient.cs (hunk 4 above)
Comment (bikamani) [reply iothub/device/src/InternalClient.cs:1067]: I actually find the previous one misleading. We also have a statement in line 1068 stating how to clear it.

File: iothub/device/src/InternalClient.cs (hunk 1 above)
Comment (bikamani) [reply iothub/device/src/InternalClient.cs:1108]: Ok sure, wasn't aware of that.

Comment (bikamani) [reply iothub/device/src/InternalClient.cs:1108]: Is it merged yet @abhipsaMisra ?

Comment (abhipsaMisra) [reply iothub/device/src/InternalClient.cs:1108]: yeah, it is in already

File: iothub/device/src/InternalClient.cs (hunk 3 above)
Comment (bikamani) [reply iothub/device/src/InternalClient.cs:112]: sure I will rename it

File: iothub/device/src/InternalClient.cs [hunk 5]
```diff
@@ -109,6 +109,7 @@ internal class InternalClient : IDisposable
         private readonly SemaphoreSlim _methodsDictionarySemaphore = new SemaphoreSlim(1, 1);
         private readonly SemaphoreSlim _deviceReceiveMessageSemaphore = new SemaphoreSlim(1, 1);
         private readonly SemaphoreSlim _moduleReceiveMessageSemaphore = new SemaphoreSlim(1, 1);
+        private readonly SemaphoreSlim _twinDesiredPropertySemaphore = new SemaphoreSlim(1, 1);
```
Comment (drwill-ms) [inline iothub/device/src/InternalClient.cs:112]: Where are these being disposed?

Comment (bikamani) [reply iothub/device/src/InternalClient.cs:112]: Taking care of this now, thanks for noticing this.

**Pull request merged**